### ISSUE TITLE
Rendering on Linux ~50% slower than on Windows

### ISSUE THREAD

**Issue by Pixelnarium** [PCSX2/pcsx2#2648]
***PCSX2 version:***
git build from today

***PCSX2 options:***
Standard Settings with OpenGL Software Mode (2 Threads are set as standard)

***Plugins used:***
GSdx AVX2

***Description of the issue:***
I am not sure what is causing this. I tried it with the game Flatout 2. As you can see on both screenshots I get around 30% on Linux and about 70% on Windows. "top" shows a CPU usage of around 180% for the pcsx2 process no matter how many threads I set. Maybe multi-threading does not work on Linux?

***PC specifications:***
Intel i5-4460 3.2Ghz, NVidia GTX 1060
Linux: ArchLinux, XFCE4, Compositor disabled
Windows: Windows 10 1803

Linux:
![pcsx2_linux](https://user-images.githubusercontent.com/30435935/47118859-10cbf080-d269-11e8-9547-e6e36d5f4707.png)

Windows:
![pcsx2_win](https://user-images.githubusercontent.com/30435935/47118862-13c6e100-d269-11e8-9de2-a8a148c954ab.png)

**Comment by Pixelnarium** [comment]
Same settings, yes "2 - safe" selection. MTVU is disabled but does not change anything. It seems like that on Windows the CPUs are better used. On Windows every core seems to be used >60% while on Linux its more like <40%. But it could just be the display.

GSdx threads are set to 4 for these pictures. 

Linux:
![pcsx2_linux](https://user-images.githubusercontent.com/30435935/47154066-912e3800-d2e1-11e8-9a25-d73e2b9ae8d0.png)

Windows:
![pcsx2_win](https://user-images.githubusercontent.com/30435935/47154072-94c1bf00-d2e1-11e8-9fc4-7270e0e1826f.png)

**Comment by mirh** [comment]
.. ok something sketchy is really going on. 
Could you try to experiment with threads number on both OS?
Also, for the records, what are numbers for hardware renderer?

**Comment by uyjulian** [comment]
Try wine with the windows version on Linux, and see how it performs

**Comment by Pixelnarium** [comment]
In WINE it runs worse. Around ~15% Performance so half of the native performance. Changing the thread numbers seems to do pretty much nothing (maybe because I only have 4 cores without HyperThreading)?. The hardware renderer is at 100% on both platforms.

**Comment by mirh** [comment]
Mhh, could you try to run some benchmark between windows and linux?
I dunno, prime95, geekbench, or anything from unigine. 

Also, try to see what happens in *fps uncapped* hw renderer.

**Comment by Pixelnarium** [comment]
Windows:
FlatOut2 Car Select OGL HW: 500% with dips down to 300%
Prime95 Benchmark 2048K 4Core/4Workers: 272 iter/sec

Linux:
FlatOut2 Car Select OGL HW: stable ~275%
Prime95 Benchmark 2048K 4Core/4Workers: 270 iter/sec

WINE:
FlatOut2 Car Select OGL HW: ~265%
Prime95 Benchmark 2048K 4Core/4Workers: 263 iter/sec

Also other software like Blender performs roughly the same between Linux and Windows.

So OK, the hardware renderer also has only 50% speed. I also tried disabling sound but that does not change anything. Any other settings I could compare? I wonder why OGL SW is so much slower in WINE when OGL HW has comparable speeds between native/WINE.

**Comment by mirh** [comment]
... starting to run out of ideas. 
How's "none" renderer instead?

**Comment by willkuer** [comment]
Not sure how it is done in pcsx2 but cross platform threading is a complex subject. Possibly the way how threads are synchronized in pcsx2 on linux performs worse than how it is done on windows.

I think turtleli worked on some pthreads topic before? Also I think this was a topic regarding the x64 port.

**Comment by Pixelnarium** [comment]
I looked more at the WINE results and saw that the wineserver process consumed up to 90% CPU time while PCSX2 renders things. The pcsx2.exe process itself consumed only ~110% so pretty much just one core. You can also see in the above CPU usage diagram that the process spend ~40% of its time inside the kernel (the orange part) for the native version. I don't know if this is "normal" or what it is doing there but it seems too much? 

PCSX2 seems to use pthreads4w. Maybe there is a settings that breaks something on Linux with the "real" pthreads?

**Comment by mirh** [comment]
>  How's "none" renderer instead?

**Comment by Farmerjoe12** [comment]
Where does threading happen in pcsx2?

**Comment by Pixelnarium** [comment]
> ... starting to run out of ideas.
> How's "none" renderer instead?

When I select "None (Core Benchmark)" it ends up in a plugin load failure
`(p) GS plugin failed to open!(thread:MTGS)(thread:EE Core)`

**Comment by Farmerjoe12** [comment]
@willkuer How is it possible that communication between threads destroys the benefits of threading? Can threads not communicate fast enough, or does the current design not support that level of communication between threads? 

Is the GS what is doing the rendering? I assume there must be some inconsistencies between windows and linux implementations if there's that much of a performance hit across both software and hardware mode. I'm not saying it's easy to have a program this complex function equivalently on vastly different operating systems, just trying to understand the problem. I'd love to take a look at where the code is thats related to this issue.

**Comment by willkuer** [comment]
Well if gdsx hw is the same as sw I don’t think it is related to threading.

Either linux just performs really poor (which I never heard before to this extend) or the regarded system is borked somehow in some sense. I only ran pcsx2 in an linux env on a vm so I can not jugde about the former. The latter is hard to find out.

With thread synchronization I meant that thread one does a job on which result thread two needs to wait to do his job. So jobs are handled sequentially on two separate threads. This is like the worst case scenario and only creates overhead. You can only efficiently multithread jobs that do not share dependencies. The more common dependencies the less efficient threading. But as I said I don’t expect it to be the current problem.

**Comment by kenshen112** [comment]
So, I'm experiencing this exact effect in reverse it's actually faster on my shitty laptop to run SW mode in Linux then it is Windows

**Comment by mirh** [comment]
How much faster?
And does opengl/direct3d sw makes any difference in windows?

**Comment by kenshen112** [comment]
@mirh There's no difference between DX SW and OGL. I get around 60 - 70% in Windows, 70 - 100% in linux without Speed hacks running KH1 as the test platform because yes my laptops that lacking.

**Comment by mirh** [comment]
Same number of threads and SSE/AVX?

**Comment by kenshen112** [comment]
AVX2 and same Number of threads

**Comment by mirh** [comment]
*sigh*... it seems #998 all over again. 
Devs are a bit busy these days, could you try experimenting with older builds (1.4.0 for example) and gcc version?

**Comment by kenshen112** [comment]
So, in 1.4 i'm getting 60 - 90%

**Comment by willkuer** [comment]
Why do you think it is related to the old skylake issue? Also apparently sometimes windows is faster sometimes linux...

**Comment by kenshen112** [comment]
Not nessacairly faster on my end it's actually a minute difference between 1.4 and 1.5 and he wasn't saying it was related but that it's kinda like that

**Comment by kenshen112** [comment]
Which in this case is leaning more towards the code

**Comment by mirh** [comment]
Could you try again null renderer? Should have been fixed.

**Comment by Pixelnarium** [comment]
Navigating to the above point using the Software plugin and then switching to the "None (Core Benchmark)" option:

Linux: ~520%
Windows: ~620%

**Comment by mirh** [comment]
That may be as well Windows hitting some other bottleneck there now... So long story short either *the whole of* this is in core, or it still is in gsdx, but core also has some inefficiency to check.

**Comment by Pixelnarium** [comment]
If threads are used in a non-parallel way switching to userspace cooperative threads (i.e. [libco](https://byuu.org/library/libco/)) could be an option to reduce the massive kernel overhead. But I am aware that this is not a simple switch to flip especially if normal threads are used in other parts.

Maybe Flatout 2 triggers some heavy overhead. I will try some other games the coming weekend.

**Comment by MrCK1** [comment]
I know it's not necessarily related to this issue but there are definitely GSdx SW bottlenecks in some places depending on the game. 

The other day, me and @RedPanda4552 tested Hitman Contracts in SW mode with 2 different CPUs (6600K vs 8700K) in the same location (on Windows). The result was that both instances ran at 23-25FPS standing in the same spot.

**Comment by kenshen112** [comment]
To be frank though, it also heavily depends on what's rendering at the moment. With KH1 there's some slowdowns depending on where I'm standing and wether or not enemies are in the scene for instance.

**Comment by Pixelnarium** [comment]
Wait.... could it just be that multithreading of GSdx "Software Mode" just does not work in Windows and "it works" in Linux but is horribly slow (at least in Flatout 2)? When I set the threads in GSdx to 0 the performance is just the same as in Windows. I did not know that setting it to 0 has any meaning (I just saw the tooltip)

**Comment by kenshen112** [comment]
@Pixelnarium that's really dependent on what's being rendered though. In kh1 go the the island there's a small fountain where you get the "Fresh" water stand in it and compare

**Comment by willkuer** [comment]
Well it‘s not like all jobs of gsdx are handled in a multithreaded way. Apparently flatout hits a path that is not multithreaded. Possibly a path with a lot of memory operations (and thus partially idling cpu).

**Comment by kenshen112** [comment]
Right but so is that fountain I mentioned

**Comment by gregory38** [comment]
The sw rasterization part of gsdx is mono thread. It could be slow on some scene.

Threading is based on c++11 and atomic. Under the hood, linux and windows have different implementation (with pros and cons).

Communication between thread is expensive. It can even be more expensive than the thread work. I think we could improve it.

**Comment by Pixelnarium** [comment]
@kenshen112 OK I see. I think you mean the waterfall in the beginning? Yeah that is like ~10% faster on Linux on both Singlethread and Multithread.

So, I don't know... if you wish you can close this issue since it seems like different operating systems really like different scenes in different games.

**Comment by mirh** [comment]
The right situation would be every OS performing "the best case" in every situation. 
So I guess like there's still something to dig up, at least until we find out the OS is the actual bottleneck, and not (say) the toolchain or whatever with libs. 

Speaking of which, #2683 would help maybe.

**Comment by MrCK1** [comment]
@Pixelnarium No change in performance recently I assume?

**Comment by arzardk** [comment]
I've done some tests today so here results.
**Linux PC**:
OS: KDE neon 5.15, Kernel: 4.18.0-18-generic
CPU: FX-6350 3.9 GHz
RAM: 16GB
GPU: GeForce 1050 Ti, driver 430.09
PCSX2 1.5.0-0 - compiled on Apr 25 2019
GS plugin: libGSdx-SSE4.so [GSdx (GCC 7.3.0 SSE4.1/AVX) 1.1.0]
**Windows PC**:
OS: Windows 10 Pro 1809
CPU: FX-6350 3.9 GHz
RAM: 16GB
GPU: GeForce 1050 Ti, driver 419.67
PCSX2-v1.5.0-dev-3083-g7f5dc2135-windows-x86
GS plugin: GSdx32-SSE4.dll [GSdx 20190424022831 (MSVC 19.00 SSE4.1/AVX) 1.1.0]
**Linux, OpenGL Software, default settings, Threads: 6**
![pcsx2-test-sw-linux2](https://user-images.githubusercontent.com/37209942/56745824-c1b8c800-6783-11e9-9c8b-422ff8c5da10.png)
**Windows, OpenGL Software, default settings, Threads: 6**
![Аннотация 2019-04-25 142349](https://user-images.githubusercontent.com/37209942/56746037-1a886080-6784-11e9-86c2-7324497c3cf1.jpg)
**Linux, OpenGL Hardware, default settings**
![pcsx2-test-hw-linux](https://user-images.githubusercontent.com/37209942/56745911-eb71ef00-6783-11e9-81e5-66598cff40ad.png)
**Windows, OpenGL Hardware, default settings**
![Аннотация 2019-04-25 142531](https://user-images.githubusercontent.com/37209942/56746639-46f0ac80-6785-11e9-9436-884016820f3a.jpg)

**Comment by willkuer** [comment]
Wouldve been nice to see the result for hw with disabled framelimiter

**Comment by mirh** [comment]
It seems pretty useless to care for that, when the frigging sw renderer is troubled to begin with. 

Can you build [this branch](https://git.kernel.org/pub/scm/linux/kernel/git/jpoimboe/linux.git/log/?h=cpu-spec-mitigations), and boot with `mitigations=off` parameter?

**Comment by Pixelnarium** [comment]
I already tested with disabled Meltdown/Spectre mitigations: https://github.com/PCSX2/pcsx2/issues/2648#issuecomment-433175791

I will do some tests again this weekend.

**Comment by gregory38** [comment]
Could be also amd arch (mt) that it is handled differently. In a less efficient way.

**Comment by gregory38** [comment]
And honestly both are slow as hell. Maybe there are too much threads.

**Comment by Pixelnarium** [comment]
My personal guess is that the FlatOut2 code triggers some bottleneck inside the Linux kernel. When you look at the CPU utilization graph above you can see that PCSX2 spends nearly half the time inside the kernel (orange part, many context switches?). Disabling multithreading in GSdx (threads = 0) solves that and does not change the performance on Windows. At least for this games. Other games benefit from more threads.

I am not sure if this is GSdx's fault here. Could also be some worst case scenario since other games are faster on Linux.

**Comment by gregory38** [comment]
Oh. Some effects require that you finish the draw before you can start a new draw. Often the draw is small enough so only 1 thread can handle it.

Potentially the game spent most of this time in the main gs thread. So nb of thread is useless.

More threads mean less work by thread. And more thread overhead.

You should first compare perf with 0 extra thread. Then 1 and 2.

**Comment by Pixelnarium** [comment]
A quick test on my notebook (i7-4750HQ 2Ghz). Can't test on Windows on this machine. 

0: 56%
1: 26%
2: 23%
3,4,5,6: 26%

Yes, 2 extra threads is a little bit slower than all other extra thread settings except 0.

**Comment by gregory38** [comment]
What likely happen is that they is more work on main thread than worker threads.
So you have this pattern
* Main send work
* Then here 2 solutions. Polling (burn cpu but less latency and can't use core for an extra work). Pit the thread to sleep (core can do others stuff but higher latency).
* Workers do the job quickly.
* Again 2 same solution for worker threads.

I'm sure you can understand which OS does what

**Comment by gregory38** [comment]
If you want to understand what happens. You need to profile with no threads to know which part of the code is busy. And potentially look at the sync point (much harder). Checking what silliness the game does in an HW renderer (Gpu debugger) might help too.

**Comment by mirh** [comment]
I'd expect profiling sw renderer would be easier?
Anyway, I linked that new tree to disable mitigations, because.. it's simply a clusterfuck to disable them otherwise (`pti=off spectre_v2=off l1tf=off nospec_store_bypass_disable` maybe?)

**Comment by Pixelnarium** [comment]
As I said, you get around 10% more performance by doing that depending on the processor. The newest Intel CPUs also have built-in fixes for Meltdown and L1TF. But that is also the case for Windows. So it does not change to underlying problem here.

**Comment by Pixelnarium** [comment]
Back on my i5-4460 3.2Ghz:

0: 73%
1,2,3,4: 30%

Maybe some interesting points while profiling PCSX2 with the perf tool (perf record). These lists are quite long so I cut them of when my screen ends:

perf report --sort=dso
Software-Rendering with 4 "extra Threads"
```
  34,50%  [kernel.kallsyms]                                                                                                                                                                    
  25,61%  libGSdx.so                                                                                                                                                                           
  15,48%  perf-936.map                                                                                                                                                                         
  13,48%  libc-2.29.so                                                                                                                                                                         
   6,98%  libpthread-2.29.so                                                                                                                                                                   
   1,04%  libspu2x-2.0.0.so                                                                                                                                                                    
   0,67%  PCSX2                                                                                                                                                                                
   0,61%  [vdso]                                                                                                                                                                               
   0,53%  libSoundTouch.so.1.0.0                                                                                                                                                               
   0,46%  libstdc++.so.6.0.25                                                                                                                                                                  
```

Software Rendering with 0 extra Threads:
```
  38,45%  perf-936.map                                                                                                                                                                         
  38,13%  libGSdx.so                                                                                                                                                                           
  14,57%  libc-2.29.so                                                                                                                                                                         
   2,87%  libspu2x-2.0.0.so                                                                                                                                                                    
   1,80%  PCSX2                                                                                                                                                                                
   0,91%  [kernel.kallsyms]                                                                                                                                                                    
   0,76%  libnvidia-glcore.so.418.56                                                                                                                                                           
   0,60%  libSoundTouch.so.1.0.0                                                                                                                                                               
   0,50%  libstdc++.so.6.0.25                                                                                                                                                                  
```

perf report
4 extra Threads:
```
   4,31%  MTGS        libc-2.29.so                   [.] _int_free                                                                                                                             
   3,87%  MTGS        libGSdx.so                     [.] GSRendererSW::Draw                                                                                                                    
   3,32%  MTGS        libc-2.29.so                   [.] _int_malloc                                                                                                                           
   2,27%  MTGS        libGSdx.so                     [.] GSRasterizer::DrawSprite                                                                                                              
   2,01%  MTGS        libGSdx.so                     [.] GSTextureCacheSW::InvalidatePages                                                                                                     
   1,90%  MTGS        [kernel.kallsyms]              [k] enqueue_entity                                                                                                                        
   1,72%  MTGS        libc-2.29.so                   [.] malloc_consolidate                                                                                                                    
   1,71%  MTGS        libGSdx.so                     [.] GSRasterizer::DrawTriangle                                                                                                            
   1,62%  MTGS        [kernel.kallsyms]              [k] _raw_spin_lock                                                                                                                        
   1,60%  MTGS        libGSdx.so                     [.] GSRendererSW::Queue                                                                                                                   
   1,48%  MTGS        libGSdx.so                     [.] GSLocalMemory::ReadTextureBlock8P                                                                                                     
   1,45%  MTGS        [kernel.kallsyms]              [k] entry_SYSENTER_compat                                                                                                                 
   1,38%  MTGS        libGSdx.so                     [.] GSRendererSW::SharedData::~SharedData                                                                                                 
   1,36%  MTGS        libpthread-2.29.so             [.] pthread_cond_wait@@GLIBC_2.3.2                                                                                                        
   1,19%  MTGS        libc-2.29.so                   [.] _mid_memalign                                                                                                                         
   1,15%  MTGS        [kernel.kallsyms]              [k] __update_load_avg_se                                                                                                                  
   1,14%  MTGS        libpthread-2.29.so             [.] __pthread_mutex_lock                                                                                                                  
   1,12%  MTGS        libpthread-2.29.so             [.] pthread_cond_signal@@GLIBC_2.3.2                                                                                                      
   1,09%  MTGS        [kernel.kallsyms]              [k] resched_curr                                                                                                                          
   1,04%  MTGS        [kernel.kallsyms]              [k] futex_wake                                                                                                                            
   1,02%  MTGS        [kernel.kallsyms]              [k] __update_load_avg_cfs_rq                                                                                                              
   1,02%  MTGS        [kernel.kallsyms]              [k] do_fast_syscall_32                                                                                                                    
   0,97%  MTGS        [kernel.kallsyms]              [k] try_to_wake_up                                                                                                                        
   0,97%  MTGS        [kernel.kallsyms]              [k] __audit_syscall_exit                                                                                                                  
   0,92%  MTGS        [kernel.kallsyms]              [k] native_write_msr                                                                                                                      
   0,87%  MTGS        libGSdx.so                     [.] GSOffset::GetPages                                                                                                                    
   0,85%  MTGS        [kernel.kallsyms]              [k] reweight_entity                                                                                                                       
   0,84%  MTGS        [kernel.kallsyms]              [k] __schedule                                                                                                                            
   0,82%  MTGS        [kernel.kallsyms]              [k] hash_futex                                                                                                                            
   0,81%  MTGS        libpthread-2.29.so             [.] __pthread_mutex_unlock_usercnt                                                                                                        
   0,77%  MTGS        libGSdx.so                     [.] GSDrawScanline::BeginDraw                                                                                                             
   0,77%  MTGS        [kernel.kallsyms]              [k] mark_wake_futex                                                                                                                       
   0,73%  MTGS        [kernel.kallsyms]              [k] update_cfs_group                                                                                                                      
   0,65%  MTGS        [kernel.kallsyms]              [k] futex_wait_setup                                                                                                                      
   0,64%  MTGS        libGSdx.so                     [.] GSRasterizer::Draw                                                                                                                    
   0,64%  MTGS        libGSdx.so                     [.] GSJobQueue<std::shared_ptr<GSRasterizerData>, 65536>::ThreadProc                                                                      
   0,63%  MTGS        [kernel.kallsyms]              [k] syscall_slow_exit_work                                                                                                                
```

0 extra Threads:
```
   5,63%  MTGS        libGSdx.so                     [.] GSRendererSW::Draw                                                                                                                    
   3,86%  MTGS        libGSdx.so                     [.] GSTextureCacheSW::InvalidatePages                                                                                                     
   3,53%  MTGS        libGSdx.so                     [.] GSRendererSW::Queue                                                                                                                   
   3,51%  MTGS        libc-2.29.so                   [.] _int_malloc                                                                                                                           
   3,05%  MTGS        libGSdx.so                     [.] GSLocalMemory::ReadTextureBlock8P                                                                                                     
   2,69%  MTGS        libc-2.29.so                   [.] _int_free                                                                                                                             
   2,44%  MTGS        libGSdx.so                     [.] GSRasterizer::DrawTriangle                                                                                                            
   2,38%  MTGS        libGSdx.so                     [.] GSRasterizer::DrawSprite                                                                                                              
   2,06%  MTGS        libc-2.29.so                   [.] malloc_consolidate                                                                                                                    
   1,88%  MTGS        libGSdx.so                     [.] GSRendererSW::SharedData::~SharedData                                                                                                 
   1,75%  EE Core     libspu2x-2.0.0.so              [.] Mix                                                                                                                                   
   1,48%  MTGS        libGSdx.so                     [.] GSLocalMemory::ReadTextureBlock32                                                                                                     
   1,21%  MTGS        libc-2.29.so                   [.] __memcpy_sse2_unaligned                                                                                                               
   1,20%  EE Core     libc-2.29.so                   [.] __memcpy_sse2_unaligned                                                                                                               
   0,83%  MTGS        libGSdx.so                     [.] GSOffset::GetPages                                                                                                                    
   0,79%  MTGS        libGSdx.so                     [.] GSLocalMemory::ReadTextureBlock16                                                                                                     
   0,75%  MTGS        libGSdx.so                     [.] GSDrawScanline::BeginDraw                                                                                                             
   0,75%  MTGS        libc-2.29.so                   [.] _mid_memalign                                                                                                                         
   0,73%  MTGS        libGSdx.so                     [.] GSLocalMemory::ReadTexture16                                                                                                          
   0,72%  MTGS        libc-2.29.so                   [.] cfree@GLIBC_2.0                                                                                                                       
   0,72%  MTGS        libc-2.29.so                   [.] _int_memalign                                                                                                                         
   0,65%  MTGS        libc-2.29.so                   [.] unlink_chunk.isra.1                                                                                                                   
   0,65%  MTGS        libGSdx.so                     [.] GSState::GIFPackedRegHandlerSTQRGBAXYZF2<4u, true>                                                                                    
   0,64%  MTGS        libGSdx.so                     [.] GSState::GetTextureMinMax                                                                                                             
   0,63%  MTGS        libGSdx.so                     [.] GSLocalMemory::WriteImage<0, 8, 8, 32>                                                                                                
   0,56%  MTGS        libGSdx.so                     [.] GSTextureCacheSW::Lookup                                                                                                              
   0,51%  MTGS        libc-2.29.so                   [.] malloc                                                                                                                                
   0,50%  MTGS        libGSdx.so                     [.] GSLocalMemory::ReadTextureBlock8HP                                                                                                    
   0,50%  MTGS        libGSdx.so                     [.] GSRasterizer::Draw                                                                                                                    
   0,49%  MTGS        libGSdx.so                     [.] GSgifTransfer                                                                                                                         
   0,47%  EE Core     libSoundTouch.so.1.0.0         [.] soundtouch::TDStretchSSE::calcCrossCorr                                                                                               
   0,47%  MTGS        libGSdx.so                     [.] std::_Hashtable<unsigned long long, std::pair<unsigned long long const, GSFunctionMap<unsigned long long, void (*)(GSVertexSW const*, 
   0,46%  MTGS        libGSdx.so                     [.] GSState::GIFRegHandlerXYZ2<6u, 0u, true>                                                                                              
   0,45%  MTGS        libGSdx.so                     [.] GSVertexTrace::Update                                                                                                                 
   0,44%  EE Core     libspu2x-2.0.0.so              [.] V_VolumeSlide::Update                                                                                                                 
   0,43%  MTGS        libGSdx.so                     [.] std::_Hashtable<unsigned long long, std::pair<unsigned long long const, GSFunctionMap<unsigned long long, void ( fastcall*)(int, int, 
   0,43%  MTGS        libGSdx.so                     [.] GSState::FlushPrim                                                                                                                    
   0,37%  EE Core     PCSX2                          [.] recEventTest                                                                                                                          

```

**Comment by arzardk** [comment]
Not sure if that is worth to mention but on GSdx (GCC 7.3.0, SSE2) 1.0.0 [libGSdx-legacy-1.0.0] performance is way better than with GSdx (GCC 7.3.0 SSE4.1/AVX) 1.1.0 [libGSdx-SSE4]. Also EE usage on the legacy GSdx (~30%) and on the new GSdx (~7%).
![Screenshot_20190427_143644](https://user-images.githubusercontent.com/37209942/56849188-0fefd780-68fa-11e9-9061-94599671e668.png)

**Comment by gregory38** [comment]
Worth to investigate auto flush (option was removed to please some people). And extra sync that I added to fix blood will tell.
Will you be able to recompile the code

**Comment by arzardk** [comment]
> Worth to investigate auto flush (option was removed to please some people). And extra sync that I added to fix blood will tell.
> Will you be able to recompile the code

I'm not familiar with how to turn on these options or where I can get the code with these options turned on. So I think I'm not be able. Sorry 😔

**Comment by Pixelnarium** [comment]
Can confirm. I built GSdx_legacy by adding "GSDX_LEGACY=TRUE" on the cmake command line. Result:

0 extra threads: 110%
4 extra threads: 150%

**Comment by gregory38** [comment]
So it is faster even with 0 extra threads ?

**Comment by Pixelnarium** [comment]
Seems so, yes. Just changed the plugin and tried it. Did not changed any settings except the extra threads.

**Comment by gregory38** [comment]
Hum. If you can edit code.
GSRendererSW.cpp. remove line 80 to 86
```
	// Reset handler with the auto flush hack enabled on the SW renderer
	// Impact on perf is rather small, and it avoids an extra hack option.
	if (!GLLoader::in_replayer) {
		m_userhacks_auto_flush = true;
		ResetHandlers();
}
```
On latest gsdx. And check again.

**Comment by Pixelnarium** [comment]
i5-4460 3.2Ghz

0: 110%
2: 155%
3: 145%
4: 130%

**Comment by gregory38** [comment]
So it is mostly on par with legacy

**Comment by soredake** [comment]
Any progress on this?

**Comment by Pixelnarium** [comment]
IIRC there is a "auto flush" checkbox in the GSDX settings that disables the things mentioned in comment https://github.com/PCSX2/pcsx2/issues/2648#issuecomment-487579694

**Comment by RedDevilus** [comment]
Can you check if this isn't fixed in the latest 1.7 dev builds? Lots of relevant work has happened in the meanwhile.

**Comment by Pixelnarium** [comment]
My old notebook broke and I have a new one. I will compare to Windows later but the "Auto Flush" Button in the Graphics Settings still reduces the performance quite significantly under Linux for the car selecting scene in Flatout 2. From 300% to 130%. I currently have v1.7.2225 installed but can update the coming weekend.

**Comment by Pixelnarium** [comment]
OK, found some time this morning. Current version v1.7.2247 on both OS. All settings the same. Last time Linux had "Edge Anti Aliasing" unchecked so I checked that too here to the setting in Windows. CPU is now a AMD Ryzen 7 5800H with power save configuration set in the EFI.

**Windows**:
0 extra threads:
- [x] auto flush: 130%
- [ ] auto flush: 170%

6 extra threads:
- [x] auto flush: 130%
- [ ] auto flush: 270%

**Linux**:
0 extra threads
- [x] auto flush: 135%
- [ ] auto flush: 180%

6 extra threads
- [x] auto flush: 140%
- [ ] auto flush: 280%

So I guess the issue is fixed? At least on my new machine here. Thank you so far :+1: This emulator really made great progress.

**Comment by tellowkrinkle** [comment]
If you're curious about what fixed it, my guess is either #4346 or #4941

#4941 because I experienced a similar discrepancy in speeds between the two OSes when I was working on it
#4346 because it's the last major change to the SW renderer and #4941 at least on macOS, #4941's changes had almost no effect if applied without #4346

If you want to test, you can disable the effects of #4346 by setting the environment variable `SPIN_WAIT_MICROSECONDS=0` (not sure how to set env variables on Windows but on Linux you can do it by just typing that on the command line in front of the path to pcsx2).  For #4941, you'll have to recompile, but I bet you can just test on master with a `git revert 6d4713e069190d6c5c2022eac72835f0cf46b57c`

For the original cause, if most of the difference is in #4346 then it's a difference in the speed of the different OSes' implementations of `std::condition_variable` and wakeups.  If it's in #4941, it's the difference in the different OSes' malloc implementations disliking the software renderer's allocation patterns more or less.

**Comment by F0bes** [comment]
Closing because there isn't much of a difference anymore.
